pkg/gadgets/trace/open/tracer/tracer_test.go

```diff
@@ -0,0 +1,216 @@
+		},
+		"test filtering": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m, err := utilstest.CreateMntNsFilterMap(info.MountNsID)
+				if err != nil {
+					testRunner.Fatalf("error creating ebpf map: %s", err)
+				}
+
+				tracer, err := NewTracer(&Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			GenerateEvent: generateEvent,
+			GetExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					MountNsID: info.MountNsID,
+					Pid:       uint32(info.Pid),
+					UID:       uint32(info.UID),
+					Comm:      info.Comm,
+					Fd:        fileFd,
+					Ret:       fileFd,
+					Err:       0,
+					Path:      "/dev/null",
+				}
+			},
+		},
+		"test UID": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m, err := utilstest.CreateMntNsFilterMap(info.MountNsID)
+				if err != nil {
+					testRunner.Fatalf("error creating ebpf map: %s", err)
+				}
+
+				tracer, err := NewTracer(&Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			RunnerConfig:  &utilstest.Config{UID: 1435},
+			GenerateEvent: generateEvent,
+			GetExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					MountNsID: info.MountNsID,
+					Pid:       uint32(info.Pid),
+					UID:       uint32(info.UID),
+					Comm:      info.Comm,
+					Fd:        fileFd,
+					Ret:       fileFd,
+					Err:       0,
+					Path:      "/dev/null",
+				}
+			},
+		},
+		"test error": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				m, err := utilstest.CreateMntNsFilterMap(info.MountNsID)
+				if err != nil {
+					testRunner.Fatalf("error creating ebpf map: %s", err)
+				}
+
+				tracer, err := NewTracer(&Config{MountnsMap: m}, nil, eventCallback)
+				if err != nil {
+					testRunner.Fatalf("error creating tracer: %s", err)
+				}
+				return tracer
+			},
+			GenerateEvent: func() error {
+				_, err := unix.Open("non-existing-file", 0, 0)
+				if err == nil {
+					return fmt.Errorf("error was expected")
+				}
+
+				return nil
+			},
+			GetExpectedEvent: func(info *utilstest.Info) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					MountNsID: info.MountNsID,
+					Pid:       uint32(info.Pid),
+					UID:       uint32(info.UID),
+					Comm:      info.Comm,
+					Fd:        0,
+					Ret:       -int(unix.ENOENT),
+					Err:       int(unix.ENOENT),
+					Path:      "non-existing-file",
+				}
+			},
+		},
+	}
+
+	testRunner.Run(tests)
```

Comment:
Instead of passing all tests to the runner, wouldn't it be better to have a simple loop here, doing some simple API calls like creating a runner and setting it up, so this code is not hidden and to make it more descriptive, like spawn runner in isolated thread or something? Now that I think about it, making tests run in parallel would play very well with having it run as a separate process instead of a separate thread.

Reply:
With the existing code, it still feels that code would be simplified if a loop would be defined on the test side, as the testing API does not make any use of running the loop, perhaps except the duplicating the loop. But I'll mark it as resolved, as it's not very important.